### ISSUE TITLE
Managing DataLoader cache

### ISSUE THREAD

**Issue by mbenedykconfigit** [graphql-dotnet/graphql-dotnet#2500]
## Summary
Is there a way of managing DataLoader cache? Oryginal facebook DataLoaders have this kind of functionality, however I was unable to find it for `qraphql-dotnet`.

## Relevant information

https://github.com/graphql/dataloader#clearing-cache

**Comment by Shane32** [comment]
The data loader has a scoped cache (per-request).  If you implement your own data loader class (inherting from `DataLoaderBase<TKey, T>`) then you can specify if the cache is enabled or disabled -- but there is no way to clear the cache or set a lifetime for objects within the cache.  However, you can implement your own global cache on top of the data loader functionality quite easily.

To implement a global cache with custom expiration options, I recommend the following:
1. Create a data loader class, inheriting `DataLoaderBase<TKey, T>` and calling the base class's constructor with caching disabled (`: base(false)`).
2. Register the class within your DI provider as a singleton
3. Include any necessary singleton services necessary for step 8c within the constructor of the class
4. Reference the data loader class from graphs in the constructor of the graph, and return `LoadAsync(key)` to request data from the data loader within your field resolvers
5. Remove all references to the data loader accessor / data loader context / data loader document listener classes.
6. Add a nuget package reference to `Microsoft.Extensions.Caching.Memory`
7. Create an instance of a `MemoryCache` within the constructor of the data loader class (or perhaps register an instance within your DI provider as a singleton, so other data loaders can use the same cache instance).
8. Override the `FetchAsync` method within the new class:
  a. Iterate through the keys, returning any from the memory cache that are possible
  b. Create a list of the remaining keys
  c. Retrieve the data for the remaining keys in a single call to your underlying data provider
  d. Save the data for those keys to the cache, setting expiration options as appropriate.

If you need to access scoped services (such as an EF `DbContext`) within step 8c, be sure to create a local service scope and pull the scoped service from that scope.

Do not attempt to use DI for the scoped service or you will run into concurrency issues.

You can register the data loader as a scoped service so that you can reference other scoped services, but then you need to move the memory cache to a static variable (or register it as a singleton), and you'll need to reference the data loader class within the field resolver via the `RequestServices` property instead of via DI.  In the end I doubt that it will make any performance difference, and the code will be messier overall.

**Comment by mbenedykconfigit** [comment]
I think current implementation of data loaders isn't flexible enough. It would be nice to have solution you described out of the box, as scenario:

>In certain uncommon cases, clearing the request cache may be necessary.
>The most common example when clearing the loader's cache is necessary is after a mutation or update within the same request, when a cached value could be out of date and future loads should not use any possibly cached value.

may be quite common, and `js` implementation have it build in.

**Comment by Shane32** [comment]
Documentation added in #2502

**Comment by Shane32** [comment]
At this time I would not add additional caching options to the data loader.  There is numerous ways to implement a cache.  Attempting to implement custom caching options introduces many questions such as:

- Should the cache be contained within the data loader instance?  Then clearing the cache would require referencing all data loaders.
- Should the cache use something specific to `GraphQL.DataLoader` or a generic implementation like `IMemoryCache`?  Using a generic implementation allows for an application-wide method of clearing the cache and controlling cache options which seems beneficial.
- If generic, what caching implementation?  Microsoft's or another?
- If specific, what cache options should be implemented?  Exposing the Microsoft classes directly and having the Microsoft dependency defeats the purpose of having a custom implementation.
- What if someone wants to use an external cache instead of a memory cache, such as a Redis cache?

Finally, implementing a custom cache is not difficult.  By overriding `LoadAsync`, the cache can be checked as each key is queued into the data loader, and by overriding `FetchAsync` it can be checked once all of the keys have been queued, so there is complete flexibility as to the implementation.  And we now have documentation describing one such implementation.

@sungam3r Do you have any thoughts here?

**Comment by sungam3r** [comment]
> There is numerous ways to implement a cache. Attempting to implement custom caching options introduces many questions.

I agree.

**Comment by Shane32** [comment]
@sungam3r  Leave open or close issue?  I propose close as a question fixed via better documentation.

**Comment by sungam3r** [comment]
OK.
